src/simsopt/mhd/vmec_diagnostics.py

```diff
@@ -986,18 +994,147 @@ def residual(theta_v, phi0, theta_p_target, jradius):
         return theta_p_target - theta_p
         """
         return theta_p_target - (theta_v + np.sum(lmns[jradius, :] * np.sin(xm * theta_v - xn * phi0)))
-
-    # Solve for theta_vmec corresponding to theta_pest:
+    
+    # A helper function that returns 0 when theta_vmec_trys are close to the
+    # proper toroidal angle in magnetic coordinates
+    # vmec_fieldlines new residual routine (written by Alan Goodman)
+    def fzero_residuals_function(theta_vmec_trys, theta_pest_targets, phis, xm, xn, lmns, lmnc, lasym):
+        angle = xm * theta_vmec_trys - xn * phis
+        fzero_residual = theta_pest_targets - (theta_vmec_trys + np.reshape(np.sum(lmns * np.sin(angle), axis=1), (-1, 1)))
+        return fzero_residual
+    
+    # A helper function that numerically finds values that minimize fzero_residuals_function
+    # This is the non-linear root finder that turns out to be ~8X faster than scipy root_scalar
+    # written by Alan Goodman
+    def get_roots(a0, b0, theta_pest_targets, phis, xm, xn, lmns, lmnc, lasym):
```

Comment:
I assumed that scipy.optimize.newton is slower than Alan's routine. Let me compare the speed of the two root finders.

Reply:
I just tested the `scipy.optimize.newton` and it's slightly faster than Alan's routine. I thought he had compared his code with the existing root finders. Here are the profiling results:
```
 Time taken to specify all arrays = 0.0372 s
 Time taken by root_scalar = 0.5243 s
 Time taken by Alan's routine = 0.0649 s
 Time taken by vectorized Newton = 0.0467 s
 Difference in theta_arrays = 2.50E-11
 Time taken by Fourier sum = 0.1972 s
 Time taken by dual relation, gradient, and all the vector calculations = 0.0010 s
 Time taken by geometric coeff calculations = 0.0013 s
```
I think I should close this PR and create a new one with the vectorized Newton solver. It would be simpler to implement and test. 

I can also separate the sign issue from this PR to prevent any confusion.

Reply:
I have removed Alan's code with `scipy.optimize.newton`
I am marking all your comments as resolved now.